### PR TITLE
Upgrading sphinx to bipass logging problems

### PR DESCRIPTION
**Pull request by john-science**
## Description

In an effort to reduce the strain of people trying to create their own ARMI-adjacent docs, this PR upgrades the version of Sphinx we are using from `2.2.0` to `4.4.0`. It also implements some important future-proofing against Python's continuous changes to the standard library method `logging.Logger.log()`. These changes together allow us to remove a bunch of calls to `logging.setLoggerClass()`, that were causing people trouble.

As part of my testing, I had GitHub Actions build these docs in Python 3.7, 3.8, and 3.9. And I verified that my changes to `runLog.py` and the docs dependencies work in all versions.

> **NOTE**:   Please do not Merge this PR!
> **NOTE**:   Please let John Merge this PR!
> **NOTE**:   John needs time to update downstream build systems before this PR is Merged!

---

## Checklist

- [X] Tests have been added/updated to verify that the new or changed code works.
- [X] Docstrings were included and updated as necessary
- [X] The code is understandable and maintainable to people beyond the author
- [X] There is no commented out code in this PR.

If user exposed functionality was added/changed:

- [X] Documentation added/updated in the `doc` folder.
- [X] New or updated dependencies have been added to `setup.py`.

### CODE REVIEW

**Comment by john-science** [comment]
@jakehader @drewejohnson This PR is in reference to Issue #518

**Comment by john-science** [comment]
@ntouran There are some minor visual differences in the resultant docs, that I could see. Hopefully I didn't miss anything. We don't really have an SOP for when we make massive changes to the docs, to make sure everyone still likes the way they look. Hmm.

I had to exactly pin the versions of these three libraries in order for the built docs to look right:

* `sphinx-gallery==0.9.0`
* `sphinx-rtd-theme==0.5.2`
* `sphinxcontrib-apidoc==0.3.0`

File: doc/requirements-docs.txt [hunk 1]
```diff
@@ -1,11 +1,30 @@
-# these are most specified that usual, because Sphinx docs seem to be quite fragile
```
Comment (ntouran) [inline doc/requirements-docs.txt:1]: I'd leave the explanation of why in the comment. that is useful.

**Review by ntouran — APPROVED** [review]
approved with one simple comment

File: doc/requirements-docs.txt (hunk 1 above)
Comment (john-science) [reply doc/requirements-docs.txt:1]: Done!

**Pull request merged**